src/query/range_query_ip_fastfield.rs

```diff
@@ -0,0 +1,197 @@
+            pos_in_data: 0,
+        }
+    }
+    fn next(&mut self) -> Option<u32> {
+        self.pos_in_data += 1;
+        self.current()
+    }
+    fn current(&self) -> Option<u32> {
+        self.data.get(self.pos_in_data).map(|el| *el as u32)
+    }
+
+    fn set_data(&mut self, data: Vec<u64>) {
+        self.data = data;
+        self.pos_in_data = 0;
+    }
+    fn is_empty(&self) -> bool {
+        self.pos_in_data >= self.data.len()
+    }
+}
+
+struct IpRangeDocSet {
+    value_range: RangeInclusive<Ipv6Addr>,
+    ip_addr_fast_field: Arc<dyn Column<Ipv6Addr>>,
+    fetched_until_doc: u32,
+    loaded_docs: VecCursor,
+}
+impl IpRangeDocSet {
+    fn new(
+        value_range: RangeInclusive<Ipv6Addr>,
+        ip_addr_fast_field: Arc<dyn Column<Ipv6Addr>>,
+    ) -> Self {
+        let mut ip_range_docset = Self {
+            value_range,
+            ip_addr_fast_field,
+            loaded_docs: VecCursor::new(),
+            fetched_until_doc: 0,
+        };
+        ip_range_docset.fetch_block();
+        ip_range_docset
+    }
+
+    /// Returns true if more data could be fetched
+    fn fetch_block(&mut self) {
+        let mut horizon: u64 = 1;
+        const MAX_HORIZON: u64 = 100_000;
+        while self.loaded_docs.is_empty() {
+            let finished_to_end = self.fetch_horizon(horizon);
+            if finished_to_end {
+                break;
+            }
+            // Fetch more data, increase horizon
+            horizon = (horizon * 2).min(MAX_HORIZON);
+        }
+    }
+
+    /// Fetches a block for docid range [fetched_until_doc .. fetched_until_doc + HORIZON]
+    fn fetch_horizon(&mut self, horizon: u64) -> bool {
+        let mut end = self.fetched_until_doc as u64 + horizon;
+        let mut finished_to_end = false;
+
+        let limit = self.ip_addr_fast_field.num_vals();
+        if end >= limit {
+            end = limit;
+            finished_to_end = true;
+        }
+
+        let data = self.ip_addr_fast_field.get_positions_for_value_range(
+            self.value_range.clone(),
+            self.fetched_until_doc as u64..end,
+        );
+        self.loaded_docs.set_data(data);
+        self.fetched_until_doc += horizon as u32;
+        finished_to_end
+    }
+}
+
+impl DocSet for IpRangeDocSet {
+    fn advance(&mut self) -> DocId {
+        if let Some(docid) = self.loaded_docs.next() {
+            docid as u32
+        } else {
+            if self.fetched_until_doc >= self.ip_addr_fast_field.num_vals() as u32 {
+                return TERMINATED;
+            }
+            self.fetch_block();
+            self.loaded_docs.current().unwrap_or(TERMINATED)
+        }
+    }
+
+    fn doc(&self) -> DocId {
+        self.loaded_docs
+            .current()
+            .map(|el| el as u32)
+            .unwrap_or(TERMINATED)
+    }
+
+    fn size_hint(&self) -> u32 {
+        0 // heuristic possible by checking number of hits when fetching a block
+    }
+}
```

Comment:
There are a lot of corner case. We need a lot of unit tests. Proptest could really help.